ludwig/utils/checkpoint_utils.py

```diff
@@ -169,42 +170,44 @@ def save(self, save_path: str, global_step: int):
           global_step (int): The iteration number which will be used
              to name the checkpoint.
         """
-        if not self.distributed.is_coordinator():
-            return
-
-        state = {
-            "global_step": global_step,
-            "model_weights": self.model.state_dict(),
-        }
-        if self.optimizer is not None:
-            state["optim_state"] = self.optimizer.state_dict()
-        if self.scheduler is not None:
-            state["scheduler_state"] = self.scheduler.state_dict()
-
-        # ignore ctrl+c while saving
-        try:
-            orig_handler = signal.getsignal(signal.SIGINT)
-            signal.signal(signal.SIGINT, lambda _sig, _frame: None)
-        except ValueError:
-            # signal throws a ValueError if we're not in the main thread
-            orig_handler = None
+        if self.is_local_rank_0():
+            state = {
+                "global_step": global_step,
+                "model_weights": self.model.state_dict(),
+            }
+            if self.optimizer is not None:
+                state["optim_state"] = self.optimizer.state_dict()
+            if self.scheduler is not None:
+                state["scheduler_state"] = self.scheduler.state_dict()
+
+            # ignore ctrl+c while saving
+            try:
+                orig_handler = signal.getsignal(signal.SIGINT)
+                signal.signal(signal.SIGINT, lambda _sig, _frame: None)
+            except ValueError:
+                # signal throws a ValueError if we're not in the main thread
+                orig_handler = None
 
-        try:
-            # atomic save
-            with tempfile.TemporaryDirectory() as tmpdir:
-                # Save to a temporary directory outside of the checkpoint dir so
-                # async processes do not try and copy a partially-written checkpoint.
-                # See Ray Tune and MLFlow for examples of background processes that
-                # are affected by this.
-                tmp_path = os.path.join(tmpdir, "temp.ckpt")
-                torch.save(state, tmp_path)
-
-                safe_move_file(tmp_path, save_path)
-                logger.debug(f"Saved checkpoint at {save_path}.")
-        finally:
-            # restore SIGINT handler
-            if orig_handler is not None:
-                signal.signal(signal.SIGINT, orig_handler)
+            try:
+                # atomic save
+                with tempfile.TemporaryDirectory() as tmpdir:
+                    # Save to a temporary directory outside of the checkpoint dir so
+                    # async processes do not try and copy a partially-written checkpoint.
+                    # See Ray Tune and MLFlow for examples of background processes that
+                    # are affected by this.
+                    tmp_path = os.path.join(tmpdir, "temp.ckpt")
+                    torch.save(state, tmp_path)
+
+                    safe_move_file(tmp_path, save_path)
```

Comment:
My understanding is that `torch.save(state, tmp_path)` will save the checkpoint to a local path on each node. My concern is that when we call `safe_move_file(tmp_path, save_path)` from every node, we'll have multiple nodes trying to write to the same path. So do we need to add an `if self.distributed.is_coordinator:` before this call?

Reply:
It's a good point, basically there are two possibilities:

1. The `save_path` is a directory local to the node (true most of the time) in which case the existing behavior works
2. The `save_path` is a remote and global filesystem like NFS, in which case this will not work

The right solution to handle (2) is to use a custom implementation of `safe_move_file` that only writes once for external filesystems. Let me give that a shot.

Reply:
I think my assumption was wrong, this happens during training so we can just save to the same path if it's local and that won't be a problem. But I think it might be a problem if it's a remote path?

Reply:
Perfect, you beat me to it :)